### DISCUSSION TITLE
MQTT topic hierarchy

### DISCUSSION THREAD

**Discussion by tisoft** [SAIC-iSmart-API/documentation#6] [General]
Proposed structure for the MQTT topic hierarchy:

- saic
    - _email_
        - messages (one subtopic for every message, currently stored on the server)
            - _messageId_
                - messageType
                - title
                - messageTime
                - sender
                - content
                - status (read/unread)
                    - set (read/unread/delete)
                - vin
        - vehicles (one subtopic for every vehicle)
            - _VIN_
                - info
                    - brand
                    - model
                    - year
                    - series
                    - color
                    - lastMessage (read only, to delete use account topic)
                        - messageId
                        - messageType
                        - title
                        - messageTime
                        - sender
                        - content
                        - status (read/unread)
                - refresh
                    - mode (off, periodic, auto)
                        - set (accepts values from above. Additionally force will trigger a single fetch, but keep the
                          current mode)
                    - lastChargeState
                    - lastVehicleState
                    - period (only used if in periodic mode)
                        - active (period in seconds to query the API, if the car is active, if not set the inActive period is used)
                            - set
                        - inActive
                            - set
                - configuration (subtopic for each entry from the modelConfigurationJsonStr field from the API)
                    - _code_
                        - name
                        - value
                - location
                    - position (GPOS position as JSON object)
                    - heading
                    - speed
                - drivetrain
                    - charging
                        - set
                    - chargerConnected
                    - chargingType (slow/fast)
                    - running
                    - auxiliaryBatteryVoltage
                    - mileage
                    - range
                    - current
                    - voltage
                    - power
                    - soc
                - climate
                    - interiorTemperature
                    - exteriorTemperature
                    - remoteTemperature (The temperature used in remote preheat commands)
                        - set
                    - remoteClimateState (off, front, on)
                        - set
                    - backWindowHeat (off, on)
                        - set
                - windows
                    - driver
                        - set
                    - passenger
                        - set
                    - rearLeft
                        - set
                    - rearRight
                        - set
                    - sunRoof
                        - set
                - doors
                    - locked
                        - set
                    - driver
                    - passenger
                    - rearLeft
                    - rearRight
                    - bonnet
                    - boot
                - tyres
                  - frontLeftPressure
                  - frontLeftPressure
                  - rearLeftPressure
                  - rearRightPressure
                - lights
                  - mainBeam
                  - dippedBeam

Topic names in _Italic_ are placeholders for the actual content. E.g. _VIN_ is replaced by the actual VIN from the
vehicle. Not all vehicles support all functionality, so some of the subtopics will not be available for each type of
vehicle.

**Comment by tisoft** [comment]
Proposal from https://github.com/SAIC-iSmart-API/saic-java-client/pull/14#issuecomment-1550404404:

Add a new topic `refresh/period/afterShutdown` to configure the grace period when the car enters the inactive state. This can be used to detect a starting charge right after the car has been switched off.

**Reply by tisoft** [reply]
I would propose the name `refresh/period/inActiveDelay` or `refresh/period/inActiveGrace`.

**Reply by mschmiedel** [reply]
I like inActiveGrace, seems to be the one that comes closest to the intention...

**Reply by nanomad** [reply]
As discussed in another issue, I would suggest a separate topic for the polling rate while during `inActiveGrace`. having it equal to the `active` period may cause 12V battery drain if mis-used

**Comment by mschmiedel** [comment]
I am missing a topic for the chrgngRmngTime from the API. how about /drivetrain/remainingChargeTime?

**Reply by mschmiedel** [reply]
I would add this feature at some point in the future, would be real helpful to see remaining time...

**Reply by mschmiedel** [reply]
Added this mr as a draft first: https://github.com/SAIC-iSmart-API/saic-java-client/pull/36

Will be testing it in the following days.

**Comment by tisoft** [comment]
`remoteClimateState` is `1` for enabling the blower, without enabling the A/C.

How should we call the new state?

Suggestions:
* `blow`
* `blowing`
* `blowOnly`
* `blowingOnly`

More suggestions?

**Reply by mschmiedel** [reply]
I like blowOnly or blowerOnly. However any option that contains "blow" should do the trick to understand that no HVAC is switched on there...

**Reply by DreamT80** [reply]
I also vote for one of the *Only alternatives. It tells us that it is a subset of the A/C that only runs part of it's functionality.

"blowingOnly" I think is the best alternative. It follows the same convention as "Heating" in "rearWindowDefrosterHeating".

**Reply by tisoft** [reply]
I will use `blowingOnly`.

**Comment by tisoft** [comment]
We have several fields in the API that contain time data. I think we should agree on how to represent them in MQTT values.

Here are the possible formats I can think of:

Timestamps:
- UNIX time 1687085943 
- ISO 8601 in UTC 2023-06-18T10:59:03.023Z
- ISO 8601 with offset 20230618T125903.023+0200
- ISO 8601 with offset and timezone 20230618T125903.023+0200[Europe/Berlin]

Durations:
- seconds as integer 90
- minutes as integer 2
- minutes as float 1.5
- ISO 8601 PT1M30S

I'm open for different formats for discussion.

**Reply by tisoft** [reply]
The Java client currently sends timestamps as quoted ISO 8601 with offset and timezone, which is quite unfortunate.
My suggestion would be sending unquoted ISO 8601 with offset.

Durations in the refresh topics are currently in seconds as integer. While I personally like ISO 8601 durations more, I think we should keep durations in seconds, since they are easier to parse for external systems.

**Reply by mschmiedel** [reply]
I agree. As I mentioned in my MR, another possibility would be to just send both formats and let people decide what they need.


```
/...../myDurationTopic/iso
/...../myDurationTopic/seconds

/...../myDateTimeTopic/iso
/...../myDateTimeTopic/timestamp
```

For duration, ISO/minutes makes just little difference, as long as the duration is not "days/months" or longer, for "hours/days" minutes are good enough to read... However, if we need seconds precision, I would rather opt for seconds than float values. In this case, having both values would offer the better readability...

**Reply by tisoft** [reply]
I changed the timestamps in the Java MQTT implementation to ISO 8601 with offset `2023-07-04T12:50:27.465827392+02:00`

**Reply by mschmiedel** [reply]
Hi, I will test this out today, hopefully no breaking change... :) 

What about the remaining charge time? The actual branch seems to work fine with my car with the actual setup for the last few weeks... Only question is if we keep minutes or want to change to something else...

**Reply by mschmiedel** [reply]
-> Just tested with HomeAssistant, works like a charm, could completely delete this hack:

value_template: "{{ ((value[:-9] | replace('\"', '') | replace('T', ' ') | as_timestamp) + 3600 * now().timetuple().tm_isdst) | timestamp_local }}"

**Reply by mschmiedel** [reply]
Adding some ideas to the conversation. The refreshPeriod configs and so are in seconds, so handling all durations in seconds seems like the best option for me... What do you think?

**Reply by tisoft** [reply]
I would also tend to use seconds everywhere. If we ever need millisecond precision, we could use fractions. (But I don't see where we would need that)

**Reply by mschmiedel** [reply]
Since we released the remainingChargingTime topic, it should be reflected in the topic hierarchy. How about adding this to some path in the documentation project? Would offer to add it there, we should just agree on some format (plain MD or something else?)
